packages/typespec-client-generator-core/src/types.ts

```diff
@@ -1060,27 +1071,31 @@ export function getSdkModelPropertyTypeBase(
   });
 }
 
-export function getSdkModelPropertyType(
+function updateMultiPartInfo(
   context: TCGCContext,
   type: ModelProperty,
-  operation?: Operation
-): [SdkModelPropertyType, readonly Diagnostic[]] {
+  base: SdkBodyModelPropertyType,
+  operation?: Operation,
+  httpOperationPart?: HttpOperationPart
+): [void, readonly Diagnostic[]] {
   const diagnostics = createDiagnosticCollector();
-  const base = diagnostics.pipe(getSdkModelPropertyTypeBase(context, type, operation));
-
-  if (isSdkHttpParameter(context, type)) return getSdkHttpParameter(context, type, operation!);
   // I'm a body model property
   let operationIsMultipart = false;
   if (operation) {
     const httpOperation = getHttpOperationWithCache(context, operation);
     operationIsMultipart = Boolean(
-      httpOperation && httpOperation.parameters.body?.contentTypes.includes("multipart/form-data")
+      (httpOperation &&
+        httpOperation.parameters.body?.contentTypes.includes("multipart/form-data")) ||
+        httpOperationPart
     );
   }
   // Currently we only recognize bytes and list of bytes as potential file inputs
+  const httpPart = getHttpPart(context.program, type.type);
+  const isRawBytesArray = base.type.kind === "array" && base.type.valueType.kind === "bytes";
   const isBytesInput =
     base.type.kind === "bytes" ||
-    (base.type.kind === "array" && base.type.valueType.kind === "bytes");
+    isRawBytesArray ||
+    (httpPart !== undefined && isOrExtendsHttpFile(context.program, httpPart.type));
```

Comment:
For the preparation codes, I would recommend we separate the "single" case and "multipart" case.

For "single", file etc. is via guesswork. For "multipart", file/contentType etc. is clear from HttpPart.
I'd prefer not to mix the guesswork and the clear data back into TCGC code.

Reply:
Make sense. I make refactor and the logic shall be more clear now.